homeassistant/components/osoenergy/binary_sensor.py

```diff
@@ -0,0 +1,95 @@
+"""Support for OSO Energy binary sensors."""
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from apyosoenergyapi import OSOEnergy
+from apyosoenergyapi.helper.const import OSOEnergyBinarySensorData
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import OSOEnergyEntity
+from .const import DOMAIN
+
+
+@dataclass(frozen=True, kw_only=True)
+class OSOEnergyBinarySensorEntityDescription(BinarySensorEntityDescription):
+    """Class describing OSO Energy heater binary sensor entities."""
+
+    value_fn: Callable[[OSOEnergy], bool]
+
+
+SENSOR_TYPES: dict[str, OSOEnergyBinarySensorEntityDescription] = {
+    "power_save": OSOEnergyBinarySensorEntityDescription(
+        key="power_save",
+        translation_key="power_save",
+        value_fn=lambda entity_data: entity_data.state,
+    ),
+    "extra_energy": OSOEnergyBinarySensorEntityDescription(
+        key="extra_energy",
+        translation_key="extra_energy",
+        value_fn=lambda entity_data: entity_data.state,
+    ),
+    "heater_state": OSOEnergyBinarySensorEntityDescription(
+        key="heater_state",
+        translation_key="heater_state",
+        value_fn=lambda entity_data: entity_data.state,
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up OSO Energy binary sensor."""
+    osoenergy = hass.data[DOMAIN][entry.entry_id]
+    devices = osoenergy.session.device_list.get("binary_sensor")
+    entities = []
+    if devices:
+        for dev in devices:
+            sensor_type = dev.osoEnergyType.lower()
+            if sensor_type in SENSOR_TYPES:
+                entities.append(
+                    OSOEnergyBinarySensor(osoenergy, SENSOR_TYPES[sensor_type], dev)
+                )
+
+    async_add_entities(entities, True)
+
+
+class OSOEnergyBinarySensor(
+    OSOEnergyEntity[OSOEnergyBinarySensorData], BinarySensorEntity
+):
+    """OSO Energy Sensor Entity."""
+
+    entity_description: OSOEnergyBinarySensorEntityDescription
+
+    def __init__(
+        self,
+        instance: OSOEnergy,
+        description: OSOEnergyBinarySensorEntityDescription,
+        entity_data: OSOEnergyBinarySensorData,
+    ) -> None:
+        """Set up OSO Energy binary sensor."""
+        super().__init__(instance, entity_data)
+
+        device_id = entity_data.device_id
```

Comment:
Do you have some examples of `entity_data.device_id`? I am trying to understand if it is a valid unique id (which must contain something like a serial number - it should remain the same if the device is removed and re-added).

See here:
https://developers.home-assistant.io/docs/entity_registry_index/#unique-id

Reply:
The device id is the unique serial number provided from the api. And example is a 128 character long string with a combination of letters and numbers that does not change for the device: qo8rexopvck59d5vmx109z1jf9dko9rqaa7afyp8mcr3rhlnn3xya4i05zy3a0i3odeoo9ztu9i4wco2f2fuy8m07ciuuddxzru5hvys2fb6jw8dqisgft5y04u2zkr7